Log error when building for Windows Desktop on non-Windows.

peterhuene opened the pull request:
This PR causes the SDK to log an error (NETSDK1100) if the user attempts to
build a Windows Desktop targeted application on a non-Windows platform.

This works via any explicit or implicit (i.e. from the SDK) framework reference
to `Microsoft.WindowsDesktop.App*`; if the current platform is not Windows and
the framework reference exists, the error is logged.

Fixes dotnet/cli#11410 and dotnet/cli#10842.

Commit:
Log error when building for Windows Desktop on non-Windows.

This commit causes the SDK to log an error (NETSDK1100) if the user attempts to
build a Windows Desktop targeted application on a non-Windows platform.

This works via any explicit or implicit (i.e. from the SDK) framework reference
to `Microsoft.WindowsDesktop.App*`; if the current platform is not Windows and
the framework reference exists, the error is logged.

Additionally, this commit prevents the Windows desktop pack from being
downloaded when building a project that doesn't reference the Windows desktop
framework on non-Windows operating systems.

Fixes dotnet/cli#11410 and dotnet/cli#10842.

peterhuene commented:
Note: we already bundle the Windows Desktop SDK for non-Windows platforms, so no changes are needed to make the SDK reference work with the "friendly" error message for those platforms.

peterhuene commented:
Also, I have a forthcoming core-sdk PR that will remove the `winforms` and `wpf` bundled templates for non-Windows builds of the .NET Core SDKs.

nguerrera commented:
> Note: we already bundle the Windows Desktop SDK for non-Windows platforms, so no changes are needed to make the SDK reference work with the "friendly" error message for those platforms.

Should we be replacing it with a stub to save disk space for code that can't be used anyway?

peterhuene commented:
I think we can consider that in another change, which would be entirely core-sdk (just drop a SDK layout that has a sdk targets file that imports the .NET Core SDK and then uses the `NETSDKError` task to generate the appropriate error from this PR).

Looks like it would save us ~4 MB on disk.

livarcocc commented:
I would rather not make the setup disk unless those 4mb actually become an issue. Also, I assume 4mb is expanded?

peterhuene commented:
That's what I'm seeing uncompressed, yes.

livarcocc reviewed (approved):
We should have @dsplaisted review this, but it LGTM.

nguerrera commented:
4mb is small enough for me to not be especially concerned about the size on disk aspect. I think there may be other benefits to avoiding any possiblity of using the WindowsDesktop SDK. I think otherwise, we're still exposed to edge cases where it kicks in and behaves randomly with its current bugs around casing and slash direction, etc. that are not getting any traction around getting fixed given that the experience of cross-compiling with it is not funded for 3.0. I would prefer to not ship code that we are investing zero effort in fixing.

That said, it's still a logically separate change from blocking the framework reference.

nguerrera commented:
I think we're missing the part where we do not schedule the download of the windows desktop targeting pack in anticipation of a possible transitive reference to it. Am I missing how that happens, or do we plan to do this as a separate change. I think we want this to work for the offline cases on non-Windows in preview 6.

peterhuene commented:
Doesn't that follow `ResolveFrameworkReferences` via the `PackagesToDownload` output? If I'm misunderstanding how this works, would you (or Daniel) pointing me at the relevant part of the targets?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -95,6 +98,14 @@ protected override void ExecuteCore()
+                {
+                    Log.LogError(Strings.WindowsDesktopFrameworkRequiresWindows);
+                    continue;
+                }
```
I read this wrong the first time and now I'm not sure how it works. Studying.

nguerrera replied:
(Aside:  really think this Target needs to be renamed. This is like the 4th time I think this is what turns framework references into individual references, but that's elsewhere.)

peterhuene replied:
`MakeFrameworkReferencesWorkPlease`

nguerrera replied:
So we're looping through all the known framework references here, and I don't get how this error wouldn't trigger always on non-Windows since I would expect windowsdekwtop to be always in this list. @dsplaisted, @peterhuene  What am I missing?

On the first read, I thought this was processing the resolved framework references, not the known framework references.

dsplaisted replied:
The `frameworkReference != null` check looks like it stops the error from happening on non-Windows.

However, I think it will still download the targeting and runtime packs, which we want to avoid.

peterhuene replied:
The logic is a little confusing.  `frameworkReferenceMap` is mapping the `FrameworkReference` names to their item specs.  We're iterating the *known* frameworks and checking each to see if it is being referenced.

Thus `frameworkReference` is only non-null if it's being referenced by the project.

Actually, now that I think about it, this should probably have a check for skipping `WindowsDesktop` on non-Windows when *not referenced*.  We could also just exclude it as a known framework reference in core-sdk based on the SDK RID, perhaps.

nguerrera replied:
> frameworkReferenceMap is mapping the FrameworkReference names to their item specs

This is where I was confused, thanks.

So the map only has the direct FrameworkReferences, right? Because this runs before we can find transitive ones in the assets file? We need to error on the transitive case as well.

nguerrera replied:
> However, I think it will still download the targeting and runtime packs, which we want to avoid.

Yes, we need to avoid this. I see that below, the adding of known framework reference to download is not conditioned on the frameworkReference != null, which it can't be because we have to eagerly download all in case we get transitive references, right?

dsplaisted replied:
It will be tricky to test the transitive FrameworkReference failure, because we won't be able to build a package or a project that has a WindowsDesktop FrameworkReference on the same machine.

As is, it should [fail](https://github.com/dotnet/sdk/blob/d6749cfae26412c188e26c247b15919afccef127/src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs#L149) with NETSDK1073: The FrameworkReference 'Microsoft.WindowsDesktop.App' was not recognized.

nguerrera replied:
> As is, it should fail with NETSDK1073: The FrameworkReference 'Microsoft.WindowsDesktop.App' was not recognized.

I think it would be useful. and much clearer if the error was NETSDK1100 in transitive case as well. Bonus points if we can have an error that "blames" the package/project that pulled in WindowsDesktop, though that would be involved and something I'd definitely handle separately. If it's going to fail with 1073 as is, I can maybe live with that for now too with an issue tracking a better diagnostic. depending on how involved fixing it is.

> It will be tricky to test the transitive FrameworkReference failure, because we won't be able to build a package or a project that has a WindowsDesktop FrameworkReference on the same machine.

Ugh, good point. We should at least manually test.

peterhuene replied:
Thanks for the feedback, both.  I will investigate this.

peterhuene replied:
Perhaps we should mark the `FrameworkReference` as windows only (copying the metadata from the known framework reference) and then again check the platform when resolving targeting pack assets to handle the transitive case? Thoughts?

peterhuene replied:
Scratch that, we don't output those currently.   Still looking for a minimally invasive way to handle the transitive case.

peterhuene replied:
I've been staring at the targets and tasks for a while now and I think I'll need more time than we have for P6 to gracefully handle the transitive case.

I'm going to push up the feedback changes and proceed and I'll file an issue to improve the transitive experience.

nguerrera commented:
Yeah, I'm confused and reading the code again.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -95,6 +98,14 @@ protected override void ExecuteCore()
 
+                if (frameworkReference != null &&
+                    frameworkReference.ItemSpec.StartsWith(WindowsDesktopFramework) &&
+                    !RuntimeEnvironment.GetRuntimeIdentifier().StartsWith("win"))
```
Not related to my other comments: I think this should be RuntimeInformation.IsOSPlatform(OSPlatform.Windows))`

peterhuene replied:
I'll make that change.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -18,6 +19,8 @@ namespace Microsoft.NET.Build.Tasks
     /// </summary>
     public class ResolveFrameworkReferences : TaskBase
     {
+        private const string WindowsDesktopFramework = "Microsoft.WindowsDesktop.App";
```
Instead of hard-coding the name here, I'd suggest adding metadata to the KnownFrameworkReference specifying that it's only supported on Windows.  Until that change makes it through core-sdk, you can add it in the ResolveFrameworkReferences target:

```xml
<KnownFrameworkReference Update="Microsoft.WindowsDesktop.App" IsWindowsOnly="True" />
<KnownFrameworkReference Update="Microsoft.WindowsDesktop.App.WindowsForms" IsWindowsOnly="True" />
<KnownFrameworkReference Update="Microsoft.WindowsDesktop.App.WPF" IsWindowsOnly="True" />
```

peterhuene replied:
Great idea.  I'll make that change.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenWeWantToRequireWindowsForDesktopApps.cs:
```diff
@@ -0,0 +1,141 @@
+                .WithProjectChanges(
+                    (filename, project) =>
+                    {
+                        project.Root.Add(XElement.Parse($@"<ItemGroup><FrameworkReference Include=""{desktopFramework}"" /></ItemGroup>"));
```
There's now a FrameworkReferences property on TestProject which you can use instead.

peterhuene replied:
Thanks, I wasn't aware of that.  I'll fix.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenWeWantToRequireWindowsForDesktopApps.cs:
```diff
@@ -0,0 +1,141 @@
+                .HaveStdOutContaining(Strings.WindowsDesktopFrameworkRequiresWindows);
+        }
+
+        private TestAsset CreateWindowsDesktopSdkTestAsset(string projectName, string desktopProperty)
```
How about using `uiFramework` or something set to "WindowsForms" or "WPF" as the parameter?  I think that would be clearer than `desktopProperty`.

peterhuene replied:
Agreed.  I'll fix.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenWeWantToRequireWindowsForDesktopApps.cs:
```diff
@@ -0,0 +1,141 @@
+                .WithProjectChanges(
+                    (filename, project) =>
+                    {
+                        project.Root.Attribute("Sdk").Value = "Microsoft.NET.Sdk.WindowsDesktop";
```
Consider adding a `ProjectSdk` property to TestProject so we have a simpler way to set the SDK.

peterhuene replied:
I can also address that; there's one or two other tests I've added that does this, so I can update those at the same time to use the new property.

dsplaisted reviewed (commented):
We should make sure that the WindowsDesktop targeting and runtime packs are not downloaded on non-Windows operating systems.  Right now I think they still will be.

Suggested tests to cover this:

- Build a Framework-dependent app with an [overridden RestorePackagesPath](https://github.com/dotnet/sdk/blob/d6749cfae26412c188e26c247b15919afccef127/src/Tests/Microsoft.NET.Build.Tests/GivenFrameworkReferences.cs#L273).  Verify that no packages were downloaded.  (This tests the offline scenario.)
- Build a self-contained app with an overridden RestorePackagesPath.  Verify that if the current operating system is not Windows, no WindowsDesktop targeting or runtime packs were downloaded.

Commit:
Add test to ensure desktop targeting pack isn't downloaded on non-Windows.

This commit adds a test that ensures no targeting pack packages get restored
when building a simple console project on Unix.

peterhuene commented:
Pushed to address review feedback (sans a graceful error message for transitive references; it receives an "unknown framework" message instead).

I added a test to ensure we don't download the windows desktop targeting pack on unix for non-desktop projects.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -91,10 +92,24 @@ protected override void ExecuteCore()
+                        Log.LogError(Strings.WindowsDesktopFrameworkRequiresWindows);
+                        windowsOnlyErrorLogged = true;
+                    }
+                    continue;
```
Add a comment to the continue statement explaining that there wasn't a direct reference to this framework and we skip the rest of the loop so that we don't pre-emptively download it in case of transitive references.

peterhuene replied:
Will add.

dsplaisted reviewed (approved):
Looks good.  I do think we should also have a test similar to `It_does_not_download_desktop_targeting_packs_on_unix` which builds a self-contained app, forcing runtime pack downloads, and verifies that the WindowsDesktop runtime packs are not downloaded.

peterhuene commented:
I'll add the self-contained test as well.

Commit:
Add test to ensure Windows desktop runtime pack isn't downloaded.

This commit adds a test that publishes a self-contained console application for
win-x64 and checks to see that the runtime pack for the Windows desktop
framework wasn't downloaded on non-Windows platforms.

peterhuene commented:
I've pushed up the self-contained test.
